Dockerfile

```diff
@@ -0,0 +1,89 @@
+WORKDIR /app
+
+# All environment variables that are available to configure on this container are listed here
+# for clarity. These list the variables supported, and the default values if not overridden
+
+# These environment variables are appended to configure the Tomcat JAVA_OPTS
+ENV OMRS_JAVA_MEMORY_OPTS="-XX:NewSize=128m"
+ENV OMRS_JAVA_SERVER_OPTS="-Dfile.encoding=UTF-8 -server -Djava.security.egd=file:/dev/./urandom -Djava.awt.headless=true -Djava.awt.headlesslib=true"
+
+# These environment variables are used to create the openmrs-server.properties file, which controls how OpenMRS initializes
+ENV OMRS_CONFIG_ADD_DEMO_DATA="false"
+ENV OMRS_CONFIG_ADMIN_USER_PASSWORD="Admin123"
+ENV OMRS_CONFIG_AUTO_UPDATE_DATABASE="true"
+ENV OMRS_CONFIG_CREATE_DATABASE_USER="false"
+ENV OMRS_CONFIG_CREATE_TABLES="false"
+ENV OMRS_CONFIG_HAS_CURRENT_OPENMRS_DATABASE="true"
+ENV OMRS_CONFIG_INSTALL_METHOD="auto"
+ENV OMRS_CONFIG_MODULE_WEB_ADMIN="true"
+
+# These variables are specific to database connections
+# Supported values for OMRS_CONFIG_CONNECTION_TYPE are "mysql" and "postgresql"
+# other values are treated as MySQL
+ENV OMRS_CONFIG_CONNECTION_TYPE="mysql"
+ENV OMRS_CONFIG_CONNECTION_USERNAME="openmrs"
+ENV OMRS_CONFIG_CONNECTION_PASSWORD="openmrs"
+ENV OMRS_CONFIG_CONNECTION_SERVER="localhost"
+ENV OMRS_CONFIG_CONNECTION_DATABASE="openmrs"
+
+# These environment variables can be used to customise the database connection.
+# Their default values depend on which database you are using.
+# OMRS_CONFIG_CONNECTION_DRIVER_CLASS
+# OMRS_CONFIG_CONNECTION_PORT
+# OMRS_CONFIG_CONNECTION_ARGS
+# OMRS_CONFIG_CONNECTION_EXTRA_ARGS
+#
+# If you really need complete control, you can just set
+# OMRS_CONFIG_CONNECTION_URL to whatever the URL should be
+
+# These environment variables are meant to enable developer settings
+# OMRS_DEV_DEBUG_PORT
+
+# Additional environment variables as needed. This should match the name of the distribution supplied OpenMRS war file
+ENV OMRS_WEBAPP_NAME="openmrs"
+
+# Copy in the start-up scripts
+COPY wait-for-it.sh /usr/local/tomcat/wait-for-it.sh
+COPY startup.sh /usr/local/tomcat/startup.sh
+RUN chmod +x /usr/local/tomcat/wait-for-it.sh
+RUN chmod +x /usr/local/tomcat/startup.sh
+
+# Take some initial steps to improve overall security, size, and functionality of the container
+# This includes updating packages and removing all of the default Tomcat webapps
+RUN sed -i '/Connector port="8080"/a URIEncoding="UTF-8" relaxedPathChars="[]|" relaxedQueryChars="[]|{}^&#x5c;&#x60;&quot;&lt;&gt;"' /usr/local/tomcat/conf/server.xml
+
+# This is run at the end primarily so that all the previous steps in building the image remain cacheable
+RUN apt-get update && apt-get install -y \
+       zip dumb-init \
+    && apt-get clean && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/* \
+    && rm -rf /usr/local/tomcat/webapps/* 
+    
+# Copy the app
+COPY --from=build /app/webapp/target/openmrs.war /usr/local/tomcat/webapps/openmrs.war
```

Comment:
So the original idea of the `/openmrs/` directory structure was to have a convention for how OpenMRS distros are laid out that the Docker image would consume, so, e.g., by placing the WAR in the `/openmrs/openmrs_core` directory and then copying it over on startup, we effectively have a re-usable image that can deploy any OpenMRS instance without a deep knowledge of where things ended up in the final Docker file tree so, core goes in `/openmrs/openmrs_core`, modules go in `/openmrs/openmrs_modules`, etc. and the Docker image works out that the WAR goes in `/usr/local/tomcat/webapps` while the OMODs go in `/openmrs/data/modules`, etc. The original thought was that we could built tooling around this standardised distribution layout that could be deployed both to Docker and to bare metal.

Reply:
100%.  It also makes it really easy if someone wants to mount in a volume and replace the "distribution" with a different one, or replace/modify certain components of it, they are in a standard location.  It also means that we can always clean up the working directories, webapps, etc. and replace them with fresh instances (remove work and temp dirs, exploded war dirs, etc) as needed, and add back in the artifacts.

Reply:
Good point! As much as I agree with having a convention to store war, modules, configs, etc. under /openmrs/ dir so it's easier to build distro images and know where to put artifacts I would advise against mounting this dir and replacing components. The correct way is to build a new image and test.  If you need to adjust configuration between environments then do this via environment variables.  If you want to clean up, then drop the container and start fresh. It's okay to do it differently for development, but for production it's walking on thin ice.